File: apps/scraper/app/executor.py
import asyncio
import contextlib
import logging
import time
from dataclasses import dataclass, field
from typing import Any, Awaitable, Callable

from playwright.async_api import Browser, Playwright, async_playwright

from .config import settings
from .models import CheckRequest
from .scraper import BrowserSessionError, ScrapeTimeoutError, scrape_product

logger = logging.getLogger(__name__)

ScrapeFunc = Callable[[str, Browser, str | None, str | None], Awaitable[dict]]


class ScrapeRequestError(RuntimeError):
    def __init__(self, status_code: int, detail: str | dict):
        super().__init__(str(detail))
        self.status_code = status_code
        self.detail = detail


@dataclass
class ScrapeJobResult:
    status_code: int
    payload: dict | None = None
    detail: str | dict | None = None


@dataclass
class ScrapeJob:
    request: CheckRequest
    enqueued_at: float = field(default_factory=time.monotonic)
    future: asyncio.Future[ScrapeJobResult] = field(init=False)

    def __post_init__(self):
        loop = asyncio.get_running_loop()
        self.future = loop.create_future()


@dataclass
class BrowserWorkerState:
    index: int
    browser: Browser | None = None
    ready: bool = False
    in_flight_started_at: float | None = None
    restart_count: int = 0
    launch_failures: int = 0
    last_launch_error: str | None = None
    last_launch_error_at: float | None = None


class ScrapeExecutor:
    def __init__(
        self,
        scrape_func: ScrapeFunc = scrape_product,
        playwright_factory: Callable[[], Any] = async_playwright,
    ):
        self._scrape_func = scrape_func
        self._playwright_factory = playwright_factory
        self.queue: asyncio.Queue[ScrapeJob] = asyncio.Queue(
            maxsize=settings.scrape_queue_size
        )
        self.worker_states = [
            BrowserWorkerState(index=index) for index in range(settings.scrape_workers)
        ]
        self.worker_tasks: list[asyncio.Task[None]] = []
        self.playwright: Playwright | None = None

    async def start(self):
        self.playwright = await self._playwright_factory().start()
        for worker in self.worker_states:
            await self._ensure_browser(worker)
        self.worker_tasks = [
            asyncio.create_task(self._worker_loop(worker))
            for worker in self.worker_states
        ]
        logger.info(
            "Scraper started configured_workers=%d ready_workers=%d queue_size=%d",
            len(self.worker_states),
            self.ready_worker_count,
            self.queue.maxsize,
        )

    async def shutdown(self):
        for task in self.worker_tasks:
            task.cancel()
        for task in self.worker_tasks:
            with contextlib.suppress(asyncio.CancelledError):
                await task
        for worker in self.worker_states:
            await self._close_browser(worker)
        if self.playwright is not None:
            await self.playwright.stop()
        logger.info("Scraper shutdown complete")

    @property
    def ready_worker_count(self) -> int:
        return sum(1 for worker in self.worker_states if worker.ready)

    def health_payload(self) -> dict:
        now = time.monotonic()
        ages_ms = [
            max(0, int((now - worker.in_flight_started_at) * 1000))
            for worker in self.worker_states
            if worker.in_flight_started_at is not None
        ]
        stuck_threshold_ms = (
            settings.scrape_job_timeout_ms + settings.health_stuck_grace_ms
        )
        stuck_workers = sum(1 for age_ms in ages_ms if age_ms > stuck_threshold_ms)
        last_launch_error = None
        last_launch_error_at = -1.0
        for worker in self.worker_states:
            if (
                worker.last_launch_error is not None
                and worker.last_launch_error_at is not None
                and worker.last_launch_error_at > last_launch_error_at
            ):
                last_launch_error = worker.last_launch_error
                last_launch_error_at = worker.last_launch_error_at

        degraded = (
            self.ready_worker_count < len(self.worker_states) or stuck_workers > 0
        )
        return {
            "status": "degraded" if degraded else "ok",
            "queue_depth": self.queue.qsize(),
            "queue_capacity": self.queue.maxsize,
            "enqueue_wait_ms": settings.scrape_enqueue_wait_ms,
            "workers": len(self.worker_states),
            "in_flight": len(ages_ms),
            "oldest_in_flight_ms": max(ages_ms, default=0),
            "stuck_workers": stuck_workers,
            "browser_workers_total": len(self.worker_states),
            "browser_workers_ready": self.ready_worker_count,
            "browser_restarts": sum(worker.restart_count for worker in self.worker_states),
            "last_launch_error": last_launch_error,
        }

    async def enqueue(self, request: CheckRequest) -> ScrapeJob:
        if self.ready_worker_count == 0:
            raise ScrapeRequestError(
                503,
                self._service_detail("no_browser_workers_ready"),
            )

        job = ScrapeJob(request)
        try:
            await asyncio.wait_for(
                self.queue.put(job),
                timeout=settings.scrape_enqueue_wait_ms / 1000,
            )
        except TimeoutError as exc:
            raise ScrapeRequestError(
                503,
                self._service_detail(
                    "queue_full",
                    url=request.url,
                    queue_depth=self.queue.qsize(),
                ),
            ) from exc
        return job

    async def _worker_loop(self, worker: BrowserWorkerState):
        while True:
            job = await self.queue.get()
            try:
                logger.info(
                    "Worker %d processing scrape for url=%r queue_depth=%d",
                    worker.index,
                    job.request.url,
                    self.queue.qsize(),
                )
                worker.in_flight_started_at = time.monotonic()
                result = await self._process_job(worker, job)
                if not job.future.done():
                    job.future.set_result(result)
            except asyncio.CancelledError:
                raise
            except Exception as exc:
                logger.exception(
                    "Worker %d failed unexpectedly for url=%r",
                    worker.index,
                    job.request.url,
                )
                if not job.future.done():
                    job.future.set_result(
                        ScrapeJobResult(
                            status_code=503,
                            detail=self._service_detail(
                                "worker_crash",
                                url=job.request.url,
                                worker_index=worker.index,
                                error=str(exc),
                            ),
                        )
                    )
            finally:
                worker.in_flight_started_at = None
                self.queue.task_done()

    async def _process_job(
        self,
        worker: BrowserWorkerState,
        job: ScrapeJob,
    ) -> ScrapeJobResult:
        if not await self._ensure_browser(worker):
            return ScrapeJobResult(
                status_code=503,
                detail=self._service_detail(
                    "browser_unavailable",
                    url=job.request.url,
                    worker_index=worker.index,
                ),
            )

        try:
            async with asyncio.timeout(settings.scrape_job_timeout_ms / 1000):
                payload = await self._scrape_func(
                    job.request.url,
                    worker.browser,
                    job.request.css_selector,
                    job.request.element_fingerprint,
                )
            return ScrapeJobResult(status_code=200, payload=payload)
        except TimeoutError:
            await self._restart_browser(worker, "scrape_job_timeout")
            return ScrapeJobResult(
                status_code=504,
                detail=self._timeout_detail(
                    "scrape_job_timeout",
                    job.request.url,
                    worker.index,
                ),
            )
        except ScrapeTimeoutError as exc:
            return ScrapeJobResult(
                status_code=504,
                detail=self._timeout_detail(
                    str(exc),
                    job.request.url,
                    worker.index,
                ),
            )
        except BrowserSessionError as exc:
            await self._restart_browser(worker, str(exc))
            return ScrapeJobResult(
                status_code=503,
                detail=self._service_detail(
                    "browser_restart_required",
                    url=job.request.url,
                    worker_index=worker.index,
                    error=str(exc),
                ),
            )

    async def _ensure_browser(self, worker: BrowserWorkerState) -> bool:
        if worker.ready and worker.browser is not None:
            return True
        if self.playwright is None:
            return False

        for attempt in range(1, settings.browser_restart_attempts + 1):
            try:
                worker.browser = await asyncio.wait_for(
                    self.playwright.chromium.launch(headless=True),
                    timeout=settings.browser_launch_timeout_ms / 1000,
                )
                worker.ready = True
                worker.launch_failures = 0
                worker.last_launch_error = None
                worker.last_launch_error_at = None
                logger.info(
                    "Browser worker %d ready on launch attempt %d",
                    worker.index,
                    attempt,
                )
                return True
            except Exception as exc:
                worker.browser = None
                worker.ready = False
                worker.launch_failures += 1
                worker.last_launch_error = str(exc)
                worker.last_launch_error_at = time.monotonic()
                logger.exception(
                    "Browser worker %d launch failed on attempt %d",
                    worker.index,
                    attempt,
                )
                if attempt < settings.browser_restart_attempts:
                    await asyncio.sleep(settings.browser_restart_backoff_ms / 1000)

        return False

    async def _restart_browser(self, worker: BrowserWorkerState, reason: str):
        worker.restart_count += 1
        logger.warning(
            "Restarting browser worker %d due to %s",
            worker.index,
            reason,
        )
        await self._close_browser(worker)
        await self._ensure_browser(worker)

    async def _close_browser(self, worker: BrowserWorkerState):
        worker.ready = False
        if worker.browser is None:
            return
        with contextlib.suppress(Exception):
            await worker.browser.close()
        worker.browser = None

    def _service_detail(self, reason: str, **extra: Any) -> dict:
        payload = self.health_payload()
        payload.update(
            {
                "reason": reason,
                **extra,
            }
        )
        return payload

    def _timeout_detail(self, reason: str, url: str, worker_index: int) -> dict:
        return {
            "status": "timeout",
            "reason": reason,
            "url": url,
            "worker_index": worker_index,
            "timeout_ms": settings.scrape_job_timeout_ms,
        }

File: apps/scraper-rs/src/executor.rs
use std::{
    sync::{
        Arc,
        atomic::{AtomicUsize, Ordering},
    },
    time::{Duration, Instant},
};

use axum::http::StatusCode;
use serde_json::{Value, json};
use tokio::{
    sync::{Mutex, mpsc, oneshot, watch},
    task::JoinHandle,
    time::timeout,
};
use tracing::{info, warn};

use crate::{
    config::Settings,
    models::{CheckRequest, CheckResponse, HealthPayload},
    runner::{BrowserHealth, ScrapeFailure, ScrapeWorker, ScrapeWorkerFactory},
};

#[derive(Debug)]
pub struct ScrapeRequestError {
    pub status_code: StatusCode,
    pub detail: Value,
}

#[derive(Debug, Clone)]
pub struct ScrapeJobResult {
    pub status_code: StatusCode,
    pub payload: Option<CheckResponse>,
    pub detail: Option<Value>,
}

struct ScrapeJob {
    request: CheckRequest,
    enqueued_at: Instant,
    response_tx: oneshot::Sender<ScrapeJobResult>,
}

pub struct ScrapeJobHandle {
    response_rx: oneshot::Receiver<ScrapeJobResult>,
}

impl ScrapeJobHandle {
    pub async fn wait(self) -> ScrapeJobResult {
        self.response_rx.await.unwrap_or(ScrapeJobResult {
            status_code: StatusCode::SERVICE_UNAVAILABLE,
            payload: None,
            detail: Some(json!({
                "status": "degraded",
                "reason": "worker_crash",
            })),
        })
    }
}

struct WorkerState {
    index: usize,
    browser_enabled: bool,
    ready: bool,
    restart_count: usize,
    last_launch_error: Option<String>,
    last_browser_error: Option<String>,
    in_flight_started_at: Option<Instant>,
}

pub struct ScrapeExecutor {
    settings: Arc<Settings>,
    worker_factory: Arc<dyn ScrapeWorkerFactory>,
    sender: mpsc::Sender<ScrapeJob>,
    receiver: Arc<Mutex<mpsc::Receiver<ScrapeJob>>>,
    shutdown_tx: watch::Sender<bool>,
    queue_depth: Arc<AtomicUsize>,
    worker_states: Vec<Arc<Mutex<WorkerState>>>,
    worker_tasks: Mutex<Vec<JoinHandle<()>>>,
}

impl ScrapeExecutor {
    pub fn new(settings: Arc<Settings>, worker_factory: Arc<dyn ScrapeWorkerFactory>) -> Self {
        let (sender, receiver) = mpsc::channel(settings.scrape_queue_size);
        let (shutdown_tx, _) = watch::channel(false);
        let worker_states = (0..settings.scrape_workers)
            .map(|index| {
                Arc::new(Mutex::new(WorkerState {
                    index,
                    browser_enabled: index < settings.browser_concurrency,
                    ready: index >= settings.browser_concurrency,
                    restart_count: 0,
                    last_launch_error: None,
                    last_browser_error: None,
                    in_flight_started_at: None,
                }))
            })
            .collect();

        Self {
            settings,
            worker_factory,
            sender,
            receiver: Arc::new(Mutex::new(receiver)),
            shutdown_tx,
            queue_depth: Arc::new(AtomicUsize::new(0)),
            worker_states,
            worker_tasks: Mutex::new(Vec::new()),
        }
    }

    pub async fn start(&self) -> Result<(), ScrapeRequestError> {
        let mut worker_tasks = self.worker_tasks.lock().await;
        if !worker_tasks.is_empty() {
            return Ok(());
        }

        let mut startup_rxs = Vec::new();
        for worker_state in &self.worker_states {
            let receiver = self.receiver.clone();
            let worker_factory = self.worker_factory.clone();
            let settings = self.settings.clone();
            let queue_depth = self.queue_depth.clone();
            let state = worker_state.clone();
            let mut shutdown_rx = self.shutdown_tx.subscribe();
            let (startup_tx, startup_rx) = oneshot::channel();
            startup_rxs.push(startup_rx);

            worker_tasks.push(tokio::spawn(async move {
                let index = { state.lock().await.index };
                let mut worker = worker_factory.create(index);
                if let Err(failure) = worker.start().await {
                    warn!(
                        worker_index = index,
                        error = %scrape_failure_detail(&failure),
                        "worker browser startup failed"
                    );
                }
                sync_worker_health(&state, worker.browser_health()).await;
                let _ = startup_tx.send(());

                loop {
                    let job = tokio::select! {
                        _ = shutdown_rx.changed() => None,
                        job = async {
                            let mut receiver = receiver.lock().await;
                            receiver.recv().await
                        } => job,
                    };

                    let Some(job) = job else {
                        break;
                    };

                    queue_depth.fetch_sub(1, Ordering::SeqCst);
                    {
                        let mut worker_state = state.lock().await;
                        worker_state.in_flight_started_at = Some(Instant::now());
                    }

                    info!(
                        worker_index = index,
                        url = %job.request.url,
                        queue_depth = queue_depth.load(Ordering::SeqCst),
                        waited_ms = job.enqueued_at.elapsed().as_millis() as u64,
                        "worker processing scrape"
                    );

                    let result =
                        process_job(worker.as_mut(), settings.clone(), index, &job.request).await;

                    let _ = job.response_tx.send(result);
                    sync_worker_health(&state, worker.browser_health()).await;

                    let mut worker_state = state.lock().await;
                    worker_state.in_flight_started_at = None;
                }

                worker.shutdown().await;
                sync_worker_health(&state, worker.browser_health()).await;
            }));
        }

        drop(worker_tasks);
        for startup_rx in startup_rxs {
            let _ = startup_rx.await;
        }

        info!(
            configured_workers = self.worker_states.len(),
            queue_size = self.settings.scrape_queue_size,
            browser_workers = self
                .settings
                .browser_concurrency
                .min(self.settings.scrape_workers),
            "rust scraper executor started"
        );

        Ok(())
    }

    pub async fn shutdown(&self) {
        let _ = self.shutdown_tx.send(true);
        let mut worker_tasks = self.worker_tasks.lock().await;
        for task in worker_tasks.drain(..) {
            let _ = task.await;
        }
    }

    pub async fn enqueue(
        &self,
        request: CheckRequest,
    ) -> Result<ScrapeJobHandle, ScrapeRequestError> {
        let (response_tx, response_rx) = oneshot::channel();
        let job = ScrapeJob {
            request: request.clone(),
            enqueued_at: Instant::now(),
            response_tx,
        };

        timeout(
            Duration::from_millis(self.settings.scrape_enqueue_wait_ms),
            self.sender.send(job),
        )
        .await
        .map_err(|_| ScrapeRequestError {
            status_code: StatusCode::SERVICE_UNAVAILABLE,
            detail: self.service_detail(
                "queue_full",
                Some(request.url.as_str()),
                Some(self.queue_depth.load(Ordering::SeqCst)),
                None,
            ),
        })?
        .map_err(|_| ScrapeRequestError {
            status_code: StatusCode::SERVICE_UNAVAILABLE,
            detail: self.service_detail(
                "executor_closed",
                Some(request.url.as_str()),
                Some(self.queue_depth.load(Ordering::SeqCst)),
                None,
            ),
        })?;

        self.queue_depth.fetch_add(1, Ordering::SeqCst);
        Ok(ScrapeJobHandle { response_rx })
    }

    pub fn health_payload(&self) -> HealthPayload {
        let now = Instant::now();
        let mut ages_ms = Vec::new();
        let mut browser_workers_total = 0;
        let mut browser_workers_ready = 0;
        let mut browser_restarts = 0;
        let mut last_launch_error = None;
        let mut last_browser_error = None;

        for worker_state in &self.worker_states {
            if let Ok(worker) = worker_state.try_lock() {
                if let Some(started_at) = worker.in_flight_started_at {
                    ages_ms.push(now.saturating_duration_since(started_at).as_millis() as u64);
                }
                if worker.browser_enabled {
                    browser_workers_total += 1;
                    if worker.ready {
                        browser_workers_ready += 1;
                    }
                }
                browser_restarts += worker.restart_count;
                if worker.last_launch_error.is_some() {
                    last_launch_error = worker.last_launch_error.clone();
                }
                if worker.last_browser_error.is_some() {
                    last_browser_error = worker.last_browser_error.clone();
                }
            }
        }

        let stuck_threshold_ms =
            self.settings.scrape_job_timeout_ms + self.settings.health_stuck_grace_ms;
        let stuck_workers = ages_ms
            .iter()
            .filter(|age_ms| **age_ms > stuck_threshold_ms)
            .count();
        let degraded = stuck_workers > 0
            || (browser_workers_total > 0 && browser_workers_ready < browser_workers_total);

        HealthPayload {
            status: if degraded {
                "degraded".to_string()
            } else {
                "ok".to_string()
            },
            queue_depth: self.queue_depth.load(Ordering::SeqCst),
            queue_capacity: self.settings.scrape_queue_size,
            enqueue_wait_ms: self.settings.scrape_enqueue_wait_ms,
            workers: self.worker_states.len(),
            in_flight: ages_ms.len(),
            oldest_in_flight_ms: ages_ms.into_iter().max().unwrap_or_default(),
            stuck_workers,
            browser_workers_total,
            browser_workers_ready,
            browser_restarts,
            last_launch_error,
            last_browser_error,
        }
    }

    fn service_detail(
        &self,
        reason: &str,
        url: Option<&str>,
        queue_depth: Option<usize>,
        error: Option<String>,
    ) -> Value {
        let mut detail = json!({
            "status": "degraded",
            "reason": reason,
        });
        if let Some(url) = url {
            detail["url"] = json!(url);
        }
        if let Some(queue_depth) = queue_depth {
            detail["queue_depth"] = json!(queue_depth);
        }
        if let Some(error) = error {
            detail["error"] = json!(error);
        }
        detail
    }
}

async fn sync_worker_health(state: &Arc<Mutex<WorkerState>>, health: BrowserHealth) {
    let mut worker_state = state.lock().await;
    worker_state.ready = if worker_state.browser_enabled {
        health.ready_workers > 0
    } else {
        true
    };
    worker_state.restart_count = health.restart_count;
    worker_state.last_launch_error = health.last_launch_error;
    worker_state.last_browser_error = health.last_browser_error;
}

async fn process_job(
    worker: &mut dyn ScrapeWorker,
    settings: Arc<Settings>,
    worker_index: usize,
    request: &CheckRequest,
) -> ScrapeJobResult {
    let timed = timeout(
        Duration::from_millis(settings.scrape_job_timeout_ms),
        worker.scrape(request),
    )
    .await;

    match timed {
        Ok(Ok(payload)) => ScrapeJobResult {
            status_code: StatusCode::OK,
            payload: Some(payload),
            detail: None,
        },
        Ok(Err(ScrapeFailure::Timeout(detail))) => ScrapeJobResult {
            status_code: StatusCode::GATEWAY_TIMEOUT,
            payload: None,
            detail: Some(timeout_detail(&detail, request.url.as_str(), worker_index)),
        },
        Ok(Err(ScrapeFailure::BrowserSession(detail))) => ScrapeJobResult {
            status_code: StatusCode::SERVICE_UNAVAILABLE,
            payload: None,
            detail: Some(service_detail(
                "browser_restart_required",
                request.url.as_str(),
                worker_index,
                Some(detail),
            )),
        },
        Err(_) => {
            warn!(worker_index = worker_index, url = %request.url, "scrape job timed out");
            worker.recycle_browser("scrape_job_timeout").await;
            ScrapeJobResult {
                status_code: StatusCode::GATEWAY_TIMEOUT,
                payload: None,
                detail: Some(timeout_detail(
                    "scrape_job_timeout",
                    request.url.as_str(),
                    worker_index,
                )),
            }
        }
    }
}

fn service_detail(reason: &str, url: &str, worker_index: usize, error: Option<String>) -> Value {
    let mut detail = json!({
        "status": "degraded",
        "reason": reason,
        "url": url,
        "worker_index": worker_index,
    });
    if let Some(error) = error {
        detail["error"] = json!(error);
    }
    detail
}

fn timeout_detail(reason: &str, url: &str, worker_index: usize) -> Value {
    json!({
        "status": "timeout",
        "reason": reason,
        "url": url,
        "worker_index": worker_index,
    })
}

fn scrape_failure_detail(failure: &ScrapeFailure) -> &str {
    match failure {
        ScrapeFailure::Timeout(detail) | ScrapeFailure::BrowserSession(detail) => detail,
    }
}
